pychromecast/controllers/spotify.py

```diff
@@ -0,0 +1,169 @@
+        """ Gets access token after login has been successful. """
+        headers = {'user-agent': USER_AGENT}
+
+        response = session.get("https://open.spotify.com/browse",
+                               headers=headers, cookies=cookies)
+        response.raise_for_status()
+
+        self.token = response.cookies['wp_access_token']
+
+        expiration = response.cookies['wp_expiration']
+        self.expiration_date = int(expiration)//1000
+
+    # Access token provided by spotipy does not have enough permissions to start
+    # Spotify on Chromecast
+    def start_session(self):
+        """ Starts session to get access token. """
+
+        # arbitrary value and can be static
+        cookies = {"__bon": "MHwwfC01ODc4MjExMzJ8LTI0Njg4NDg3NTQ0fDF8MXwxfDE="}
+
+        if self.username is None:
+            self.username = os.getenv("SPOTIFY_USERNAME")
+
+        if self.password is None:
+            self.password = os.getenv("SPOTIFY_PASS")
+
+        if self.username is None or self.password is None:
+            raise Exception("No username or password")
+
+        session = requests.Session()
+        token = self. _get_csrf(session, cookies)
+
+        self._login(session, cookies, self.username, self.password, token)
+        self._get_access_token(session, cookies)
+
+        self.session_started = True
+
+        return self.token
+
+    def launch_app(self):
+        """ Launch main application """
+
+        def callback():
+            """Callback function"""
+            self.send_message({"type": TYPE_STATUS, "credentials": self.token})
+
+        curr_time = time.time()
+
+        if self.session_started and curr_time < self.expiration_date:
+            self.logger.debug("Using same token: %s", self.token)
+            self.launch(callback_function=callback)
+        else:
+            self.logger.debug("Creating new token")
+            self.start_session()
+            self.logger.debug("Token is: %s", self.token)
+            self.launch(callback_function=callback)
+
+        # Need to wait for Spotify to be launched on Chromecast completely
+        time.sleep(5)
+        self.client = spotipy.Spotify(auth=self.token)
+
+        self.device_id = self.get_spotify_device_id()
+
+    def get_spotify_device_id(self):
+        """
+        Gets device id from Spotify. This should be the same as the name of
+        the Chromecast.
+        """
+        devices_available = self.client.devices()
+
+        for device in devices_available['devices']:
+            self.logger.debug(device)
+            if device['name'] == self.cast_name:
+                return device['id']
+        return None
+
+    def play_song(self, uri):
+        """ Play a single song with it's Spotify URI. """
+        if self.device_id is None:
+            raise Exception("No device id. Try launching app again")
+        else:
+            self.client.start_playback(device_id=self.device_id, uris=[uri])
```

Comment:
Personally, I would unindent this code block as above expression already `raise`s. See http://blog.timoxley.com/post/47041269194/avoid-else-return-early